customreboot/grub.js

```diff
@@ -0,0 +1,205 @@
+      if (cfgpath == "") {
+          throw new String("Failed to find grub config");
+      }
+
+      let bootOptions = new Map();
+
+      let defualtEn = "";
+
+      let file = Gio.file_new_for_path(cfgpath);
+      let [suc, content] = file.load_contents(null);
+      if (!suc) {
+          throw new String("Failed to load grub config");
+      }
+
+      let lines;
+      if (content instanceof Uint8Array) {
+          lines = new TextDecoder().decode(content);
+      }
+      else {
+          lines = content.toString();
+      }
+
+      let entryRx = /^menuentry ['"]([^'"]+)/;
+      let defaultRx = /(?<=set default=\")([A-Za-z- ()/0-9]*)(?=\")/
+      lines.split('\n').forEach(l => {
+          let res = entryRx.exec(l);
+          if (res && res.length) {
+              bootOptions.set(res[1], res[1]);
+          }
+          let def = defaultRx.exec(l);
+          if (def && def.length) {
+              defualtEn = def[0];
+          }
+      });
+
+      bootOptions.forEach((v, k) => {
+          Log(`${k} = ${v}`);
+      });
+
+      if (defualtEn == "") defualtEn = bootOptions.keys().next().value;
+
+      return [bootOptions, defualtEn];
+          
+    } catch (e) {
+      LogWarning(e);
+      return undefined;
+    }
+  }
+
+  /**
+   * Set's the next boot option
+   * @param {string} id 
+   * @returns True if the boot option was set, otherwise false
+   */
+  static async SetBootOption(id) {
+    try {
+      let [status, stdout, stderr] = await ExecCommand(
+          ['/usr/bin/pkexec', '/usr/sbin/grub-reboot', id],
+      );
+      Log(`Set boot option to ${id}: ${status}\n${stdout}\n${stderr}`);
+      return true;
+    } catch (e) {
+      LogWarning(e);
+      return false;
+    }
+  }
+
+  /**
+   * Can we use this bootloader?
+   * @returns True if useable otherwise false
+   */
+  static async IsUseable() {
+    return await this.GetConfig() !== "";
+  }
+
+  /**
+   * Get's grub config file
+   * @returns A string containing the location of the config file, if none is found returns a blank string
+   */
+  static async GetConfig() {
+    let paths = ["/boot/grub/grub.cfg", "/boot/grub2/grub.cfg"];
+
+    let file;
+
+    for (let i = 0; i < paths.length; i++) {
+        file = Gio.file_new_for_path(paths[i]);
+        if (file.query_exists(null)) {
+            return paths[i];
+        }
+    }
+
+    return "";
+  }
+
+  /**
+   * Copies a custom grub script to allow the extension to quickly reboot into another OS
+   * If anyone reads this: Idk how to combine these into one pkexec call, if you do please leave a commit fixing it
+   */
+  static async EnableQuickReboot(ext) {
+    try {
+      let [status, stdout, stderr] = await ExecCommand([
+          'pkexec',
+          'sh',
+          '-c',
+          `/usr/bin/cp ${Extension.lookupByUUID('hibernate-status@dromi').path}/customreboot/42_custom_reboot /etc/grub.d/42_custom_reboot && /usr/bin/chmod 755 /etc/grub.d/42_custom_reboot && /usr/sbin/update-grub`
```

Comment:
The `update-grub` command has been replaced by `grub-mkconfig`: https://git.savannah.gnu.org/cgit/grub.git/tree/NEWS#n537

Reply:
I'll change this, thanks.